Fix mangling of '@noDerivative' and unify function attribute mangling operators.

rxwei opened the pull request:
* Unify mangling operators for attributes `async`, `@Sendable`, and `@differentiable` under a single `Y` operator.  Free up operators `J` and `j`.
    ```
    async ::= 'Ya'                             // 'async' annotation on function types
    sendable ::= 'Yb'                          // @Sendable on function types
    throws ::= 'K'                             // 'throws' annotation on function types
    differentiable ::= 'Yjf'                   // @differentiable(_forward) on function type
    differentiable ::= 'Yjr'                   // @differentiable(reverse) on function type
    differentiable ::= 'Yjd'                   // @differentiable on function type
    differentiable ::= 'Yjl'                   // @differentiable(_linear) on function type
    ```

* `@noDerivative` was not mangled in function types, and was resolved incorrectly when there's an ownership specifier. It is fixed by this patch with the following changes:
    * Add `NoDerivative` demangle node represented by a `Yk` operator.
        ```
        list-type ::= type identifier? 'Yk'? 'z'? 'h'? 'n'? 'd'?  // type with optional label, '@noDerivative', inout convention, shared convention, owned convention, and variadic specifier
        ```
    * Fix `NoDerivative`'s overflown offset in `ParameterTypeFlags` (`7` -> `6`).
    * In type decoder and type resolver where attributed type nodes are processed, add support for nested attributed nodes, e.g. `inout @noDerivative T`.
    * Add `TypeResolverContext::InoutFunctionInput` so that when we resolve an `inout @noDerivative T` parameter, the `@noDerivative T` checking logic won't get a `TypeResolverContext::None` set by the caller.

Resolves rdar://76299796 and rdar://75916833.

rxwei committed:
[AutoDiff] Fix mangling of '@noDerivative' in function types.

`@noDerivative` was not mangled in function types, and was resolved incorrectly when there's an ownership specifier. It is fixed by this patch with the following changes:

* Add `NoDerivative` demangle node represented by a `k` operator.
    ```
    list-type ::= type identifier? 'k'? 'z'? 'h'? 'n'? 'd'?  // type with optional label, '@noDerivative', inout convention, shared convention, owned convention, and variadic specifier
    ```
* Fix `NoDerivative`'s overflown offset in `ParameterTypeFlags` (`7` -> `6`).
* In type decoder and type resolver where attributed type nodes are processed, add support for nested attributed nodes, e.g. `inout @noDerivative T`.
* Add `TypeResolverContext::InoutFunctionInput` so that when we resolve an `inout @noDerivative T` parameter, the `@noDerivative T` checking logic won't get a `TypeResolverContext::None` set by the caller.

Resolves rdar://75916833.

rxwei commented:
@swift-ci please test

joshnewnham reviewed (approved):
LGTM

eeckstein reviewed (changes requested):
Can you pease not use a top-level operator ('k') for this?
We only have very few left and we should not "waste" them for something which is not critical for symbol sizes.

rxwei commented:
> Can you pease not use a top-level operator ('k') for this?
> We only have very few left and we should not "waste" them for something which is not critical for symbol sizes.

Could you suggest an existing operator that I fold under?

rxwei commented:
Should the same rationale apply to `AsyncAnnotation` ('Y')?  I could make `Y` be an umbrella operator for annotations/markers if you think that is appropriate, the `Ya` could be `async` and `Yj` could be `@noDerivative`.

Similarly, `J` is currently used only for `Node::Kind::ConcurrentFunctionType`, which is an annotation just like `Node::Kind::DifferentiableFunctionType`. I can move both of them under  `Y`/`J` if there are no objections.

rxwei committed:
Unify mangling operators for `async`, `@Sendable`, `@differentiable` and `@noDerivative`.

Repurpose mangling operator `Y` as an umbrella operator that covers new attributes on function types. Free up operators `J`, `j`, and `k`.

```
async ::= 'Ya'                             // 'async' annotation on function types
sendable ::= 'Yb'                          // @Sendable on function types
throws ::= 'K'                             // 'throws' annotation on function types
differentiable ::= 'Yjf'                   // @differentiable(_forward) on function type
differentiable ::= 'Yjr'                   // @differentiable(reverse) on function type
differentiable ::= 'Yjd'                   // @differentiable on function type
differentiable ::= 'Yjl'                   // @differentiable(_linear) on function type
```

Resolves rdar://76299796.

rxwei commented:
In 3c5b367 I made the change I proposed above. PTAL.

```
async ::= 'Ya'                             // 'async' annotation on function types
sendable ::= 'Yb'                          // @Sendable on function types
throws ::= 'K'                             // 'throws' annotation on function types
differentiable ::= 'Yjf'                   // @differentiable(_forward) on function type
differentiable ::= 'Yjr'                   // @differentiable(reverse) on function type
differentiable ::= 'Yjd'                   // @differentiable on function type
differentiable ::= 'Yjl'                   // @differentiable(_linear) on function type
```

DougGregor reviewed (approved):
Skimmed this, but it LGTM

rxwei commented:
@swift-ci please test

eeckstein commented on include/swift/Demangling/TypeDecoder.h:
```diff
@@ -1379,28 +1380,39 @@ class TypeDecoder {
+      while (recurse) {
+        switch (node->getKind()) {
+        case NodeKind::InOut:
+          setOwnership(ValueOwnership::InOut);
```
This will result in an infinite loop. You should set the ``recurse`` flag only to true for cases where you also change ``node``. E.g.
```
bool recurse = false;
do {
  recurse = false;
  switch (node->getKind()) {
    ...
    case NodeKind::NoDerivative:
      node = node->getFirstChild();
      recurse = true;
      break;
    ...
} while (recurse);
```
Also, can you add a test for this?

rxwei replied:
The current implementation won't case infinite loops, because `setOwnership` always changes `node` to its child.

eeckstein replied:
ah, I didn't see that ``setOwnership`` changes node.

eeckstein replied:
IMO, this is really subtle. It would be much simpler to just manually inline the few lines of ``setOwnership``.

rxwei replied:
Sounds good. I've inlined them.

eeckstein commented on lib/Demangling/Demangler.cpp:
```diff
@@ -783,7 +782,22 @@ NodePointer Demangler::demangleOperator() {
+            createWithChild(Node::Kind::NoDerivative, popTypeAndGetChild()));
+      default:
+        pushBack();
+        pushBack();
```
Why this? I think it should just return null in case there is an invalid 'Y' operator.

rxwei replied:
This is mirroring line 765 where sub-operators of 'H' are being parsed. I didn't understand why it existed there, so I copied the behavior just in case. I've removed it now since you requested it.

eeckstein replied:
yeah, also the 'H' demangling should just return a nullptr.

eeckstein commented on lib/Demangling/Demangler.cpp:
```diff
@@ -783,7 +782,22 @@ NodePointer Demangler::demangleOperator() {
     case 'X': return demangleSpecialType();
-    case 'Y': return createNode(Node::Kind::AsyncAnnotation);
+    case 'Y':
+      switch (char c2 = nextChar()) {
```
cosmetic: Can you extract this switch to a separate function? Just to keep ``demangleOperator`` readable.

rxwei replied:
Extracted to `Demangler::demangleTypeAnnotation()`.

eeckstein reviewed (changes requested):
Thanks, the mangling scheme looks good now.
I just have a few comments on the implementation.

rxwei commented:
@swift-ci please test

eeckstein reviewed (approved):
LGTM, thanks!

rxwei commented:
@swift-ci please test

rxwei commented:
@swift-ci please test and merge

rxwei commented:
@swift-ci please test and merge

rxwei commented:
@swift-ci please test macos

rxwei commented:
@swift-ci please test macos
